File: app/src/main/java/quintype/com/templatecollectionwithrx/models/entities/EntityModel.java
package quintype.com.templatecollectionwithrx.models.entities;

import android.os.Parcel;
import android.os.Parcelable;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.annotations.SerializedName;
import com.google.gson.reflect.TypeToken;

public abstract class EntityModel implements Parcelable {
    @SerializedName("updated-at")
    protected long updatedAt;
    @SerializedName("publisher-id")
    protected String publisherId;
    @SerializedName("name")
    protected String name;
    @SerializedName("type")
    protected String type;
    @SerializedName("entity-type-id")
    protected String entityTypeId;
    @SerializedName("deleted-at")
    protected long deletedAt;
    @SerializedName("created-by")
    protected String createdBy;
    @SerializedName("id")
    protected String id;
    @SerializedName("last-updated-by")
    protected String lastUpdatedBy;
    @SerializedName("created-at")
    protected String createdAt;

    public long updatedAt() {
        return this.updatedAt;
    }

    public String publisherId() {
        return this.publisherId;
    }

    public String name() {
        return this.name;
    }

    public String type() {
        return this.type;
    }

    public String entityTypeId() {
        return this.entityTypeId;
    }

    public long deletedAt() {
        return this.deletedAt;
    }

    public String createdBy() {
        return this.createdBy;
    }

    public String id() {
        return this.id;
    }

    public String lastUpdatedBy() {
        return this.lastUpdatedBy;
    }

    public String createdAt() {
        return this.createdAt;
    }

    protected EntityModel(Parcel in) {
        this.updatedAt = in.readLong();
        this.publisherId = in.readString();
        this.name = in.readString();
        this.type = in.readString();
        this.entityTypeId = in.readString();
        this.deletedAt = in.readLong();
        this.createdBy = in.readString();
        this.id = in.readString();
        this.lastUpdatedBy = in.readString();
        this.createdAt = in.readString();
    }

    public EntityModel() {
    }

    public String toString() {
        return "updatedAt=" + this.updatedAt + ", publisherId='" + this.publisherId + '\'' + ", name='" + this.name + '\'' + ", type='" + this.type + '\'' + ", entityTypeId='" + this.entityTypeId + '\'' + ", deletedAt=" + this.deletedAt + ", createdBy='" + this.createdBy + '\'' + ", id='" + this.id + '\'' + ", lastUpdatedBy='" + this.lastUpdatedBy + '\'' + ", createdAt='" + this.createdAt;
    }

    public static <ENTITY extends EntityModel> ENTITY createFromJson(JsonObject entityJson, TypeToken<ENTITY> entityType) {
        Gson gson = new Gson();
        return (ENTITY) gson.fromJson(entityJson, entityType.getType());
    }

    public static <ENTITY extends EntityModel> ENTITY createFromJson(JsonObject entityJson, Class<ENTITY> clazz) {
        Gson gson = new Gson();
        return (ENTITY) gson.fromJson(entityJson, clazz);
    }

    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(this.updatedAt);
        dest.writeString(this.publisherId);
        dest.writeString(this.name);
        dest.writeString(this.type);
        dest.writeString(this.entityTypeId);
        dest.writeLong(this.deletedAt);
        dest.writeString(this.createdBy);
        dest.writeString(this.id);
        dest.writeString(this.lastUpdatedBy);
        dest.writeString(this.createdAt);
    }
}

File: androidcore/src/main/java/com/example/androidcore/models/entities/EntityModel.kt
//package com.example.androidcore.models.entities
//
//import android.os.Parcel
//import android.os.Parcelable
//import com.google.gson.Gson
//import com.google.gson.JsonObject
//import com.google.gson.annotations.SerializedName
//import com.google.gson.reflect.TypeToken
//
///**
// * Created TemplateCollectionWithRx by rakshith on 8/31/18.
// */
//
///**
// * Basic Entity. Should not be used as is. Extend this class in your app and make it Parcelable.
// *
// * Following are the common values in all entities:
// *
// *
// * "updated-at": 1499191886946,
// * "publisher-id": 70,
// * "name": "Frenny Bawa",
// * "type": "individual",
// * "entity-type-id": 29,
// * "deleted-at": null,
// * "created-by": 88015,
// * "id": 32423,
// * "last-updated-by": 88015,
// * "created-at": 1499191886946
// */
//
//abstract class EntityModel : Parcelable {
//    @SerializedName("updated-at")
//    protected var updatedAt: Long = 0
//    @SerializedName("publisher-id")
//    protected var publisherId: String
//    @SerializedName("name")
//    protected var name: String
//    @SerializedName("type")
//    protected var type: String
//    @SerializedName("entity-type-id")
//    protected var entityTypeId: String
//    @SerializedName("deleted-at")
//    protected var deletedAt: Long = 0
//    @SerializedName("created-by")
//    protected var createdBy: String
//    @SerializedName("id")
//    public var id: String
//    @SerializedName("last-updated-by")
//    protected var lastUpdatedBy: String
//    @SerializedName("created-at")
//    protected var createdAt: String
//
//    protected constructor(parcel: Parcel) {
//        this.updatedAt = parcel.readLong()
//        this.publisherId = parcel.readString()
//        this.name = parcel.readString()
//        this.type = parcel.readString()
//        this.entityTypeId = parcel.readString()
//        this.deletedAt = parcel.readLong()
//        this.createdBy = parcel.readString()
//        this.id = parcel.readString()
//        this.lastUpdatedBy = parcel.readString()
//        this.createdAt = parcel.readString()
//    }
//
//    /**
//     * Override this method in classes that extend EntityModel and add the extra fields in this
//     * form, e.g.:
//     *
//     * return "BookEntity{" +               //Name of your class + starting bound
//     * super.toString() +            //Call super for the common fields
//     * ", link='" + link + '\'' +    //Add the extra fields
//     * '}'; //end bound
//     *
//     * @return a string with all the common fields in Entities
//     */
//    override fun toString(): String {
//        return "updatedAt=" + updatedAt +
//                ", publisherId='" + publisherId + '\''.toString() +
//                ", name='" + name + '\''.toString() +
//                ", type='" + type + '\''.toString() +
//                ", entityTypeId='" + entityTypeId + '\''.toString() +
//                ", deletedAt=" + deletedAt +
//                ", createdBy='" + createdBy + '\''.toString() +
//                ", id='" + id + '\''.toString() +
//                ", lastUpdatedBy='" + lastUpdatedBy + '\''.toString() +
//                ", createdAt='" + createdAt
//    }
//
//    override fun writeToParcel(dest: Parcel, flags: Int) {
//        dest.writeLong(this.updatedAt)
//        dest.writeString(this.publisherId)
//        dest.writeString(this.name)
//        dest.writeString(this.type)
//        dest.writeString(this.entityTypeId)
//        dest.writeLong(this.deletedAt)
//        dest.writeString(this.createdBy)
//        dest.writeString(this.id)
//        dest.writeString(this.lastUpdatedBy)
//        dest.writeString(this.createdAt)
//    }
//
//    companion object {
//
//        /**
//         * Use the methods below to get an instance of a specific entity type: one that extends this
//         * class
//         *
//         * @param entityJson the whole entity in JSON form
//         * @param entityType the Type or Class name of your entity
//         * @param <ENTITY> generic class that extends EntityModel
//         * @return the JSON data deserialised into your entity class
//        </ENTITY> */
//
//        fun <ENTITY : EntityModel> createFromJson(entityJson: JsonObject, entityType: TypeToken<ENTITY>): ENTITY? {
//            val gson = Gson()
//            return gson.fromJson<ENTITY>(entityJson, entityType.type)
//        }
//
//        fun <ENTITY : EntityModel> createFromJson(entityJson: JsonObject, clazz: Class<ENTITY>): ENTITY {
//            val gson = Gson()
//            return gson.fromJson(entityJson, clazz)
//        }
//    }
//
//    /*
//    Fixed the parcelable model according to the Stack overflow posts
//    https://stackoverflow.com/a/10841502
//    https://stackoverflow.com/a/35250396
//     */
//}
//
